sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java

```diff
@@ -0,0 +1,878 @@
+        if (currentValue != null) {
+          elementIsPresent = true;
+          startOfRecord += codec.recordLength(currentValue);
+          return true;
+        } else {
+          elementIsPresent = false;
+          return false;
+        }
+      }
+    }
+  }
+
+  /**
+   * A {@link FileBasedSink} for TFRecord files. Produces TFRecord files.
+   */
+  @VisibleForTesting
+  static class TFRecordSink extends FileBasedSink<byte[]> {
+    @VisibleForTesting
+    TFRecordSink(ValueProvider<String> baseOutputFilename,
+                 String extension,
+                 String fileNameTemplate,
+                 TFRecordIO.CompressionType compressionType) {
+      super(baseOutputFilename, extension, fileNameTemplate,
+          writableByteChannelFactory(compressionType));
+    }
+
+    @Override
+    public FileBasedWriteOperation<byte[]> createWriteOperation(PipelineOptions options) {
+      return new TFRecordWriteOperation(this);
+    }
+
+    private static WritableByteChannelFactory writableByteChannelFactory(
+        TFRecordIO.CompressionType compressionType) {
+      switch (compressionType) {
+        case AUTO:
+          throw new IllegalArgumentException("Unsupported compression type AUTO");
+        case NONE:
+          return CompressionType.UNCOMPRESSED;
+        case GZIP:
+          return CompressionType.GZIP;
+        case DEFLATE:
+          return CompressionType.DEFLATE;
+      }
+      return CompressionType.UNCOMPRESSED;
+    }
+
+    /**
+     * A {@link org.apache.beam.sdk.io.FileBasedSink.FileBasedWriteOperation
+     * FileBasedWriteOperation} for TFRecord files.
+     */
+    private static class TFRecordWriteOperation extends FileBasedWriteOperation<byte[]> {
+      private TFRecordWriteOperation(TFRecordSink sink) {
+        super(sink);
+      }
+
+      @Override
+      public FileBasedWriter<byte[]> createWriter(PipelineOptions options) throws Exception {
+        return new TFRecordWriter(this);
+      }
+    }
+
+    /**
+     * A {@link org.apache.beam.sdk.io.FileBasedSink.FileBasedWriter FileBasedWriter}
+     * for TFRecord files.
+     */
+    private static class TFRecordWriter extends FileBasedWriter<byte[]> {
+      private WritableByteChannel outChannel;
+      private TFRecordCodec codec;
+
+      private TFRecordWriter(FileBasedWriteOperation<byte[]> writeOperation) {
+        super(writeOperation);
+        this.mimeType = MimeTypes.BINARY;
+      }
+
+      @Override
+      protected void prepareWrite(WritableByteChannel channel) throws Exception {
+        this.outChannel = channel;
+        this.codec = new TFRecordCodec();
+      }
+
+      @Override
+      public void write(byte[] value) throws Exception {
+        codec.write(outChannel, value);
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  private static class TFRecordCodec {
+    private static final int HEADER_LEN = (Long.SIZE + Integer.SIZE) / Byte.SIZE;
```

Comment:
Are they (Integer.SIZE) variant to the system? Why do you calculate in this way?
The size of HEADER/FOOTER should not change per system, or else you'll fail to read the data generated by other systems.